src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlBinaryWriter.cs

```diff
@@ -870,15 +875,146 @@ private void WriteArrayInfo(XmlBinaryNodeType nodeType, int count)
             WriteMultiByteInt32(count);
         }
 
-        public unsafe void UnsafeWriteArray(XmlBinaryNodeType nodeType, int count, byte* array, byte* arrayMax)
+        public unsafe void UnsafeWriteBoolArray(bool[] array, int offset, int count)
+        {
+            WriteArrayInfo(XmlBinaryNodeType.BoolTextWithEndElement, count);
+            fixed (bool* items = &array[offset])
+            {
+                base.UnsafeWriteBytes((byte*)items, count);
+            }
+        }
+
+        public unsafe void UnsafeWriteInt16Array(short[] array, int offset, int count)
+        {
+            WriteArrayInfo(XmlBinaryNodeType.Int16TextWithEndElement, count);
+            if (BitConverter.IsLittleEndian)
+            {
+                fixed (short* items = &array[offset])
+                {
+                    base.UnsafeWriteBytes((byte*)items, 2 * count);
+                }
+            }
+            else
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    Span<byte> span = GetBuffer(2, out int bufferOffset).AsSpan(bufferOffset, 2);
+                    BinaryPrimitives.WriteInt16LittleEndian(span, array[offset + i]);
+                    Advance(2);
+                }
+            }
+        }
+
+        public unsafe void UnsafeWriteInt32Array(int[] array, int offset, int count)
+        {
+            WriteArrayInfo(XmlBinaryNodeType.Int32TextWithEndElement, count);
+            if (BitConverter.IsLittleEndian)
+            {
+                fixed (int* items = &array[offset])
+                {
+                    base.UnsafeWriteBytes((byte*)items, 4 * count);
+                }
```

Comment:
We might want to consider doing a similar approach for little endian as we do for bin endian as UnsafeWriteBytes forces a buffer flush whereas the big endian implementation only causes a buffer flush when the buffer doesn't have space.  
We could also improve the performance of the big endian implementation by having a faster path when the size of the array is smaller than the buffer length. In that case you can call GetBuffer once and then have the loop wrap the write call, then call Advance once at the end. Could also do this for bigger arrays and just do it in batches which are smaller than the buffer size. The buffer length is hard coded to 512 bytes so these could be done in batches of 128 count.  
None of that is required for this PR, just noting improvements that I can see which are possible.

Reply:
I had considered that. But for a very late-stage PR in 7.0 that we can hopefully bring back to 6.0... I thought it's probably better to stay faithful to the existing little-endian implementation. As you say, it's not required for this PR.

Reply:
Agreed.  We can do performance improvements for big-endian systems for .NET 8.   I'd be happy to work on this.